components/cam/src/physics/cam/shoc.F90

```diff
@@ -682,6 +677,70 @@ subroutine update_prognostics( &
 
 end subroutine update_prognostics
 
+!==============================================================
+! Update tracers based on SGS mixing
+! using explicit diffusion solver.  Note that this routine
+! should only be called if using very small time steps
+! (< 20 s), otherwise the implicit diffusion solver should be used
+
+subroutine update_tracers( &
+         shcol,nlev,nlevi,num_tracer,&    ! Input
+         dtime,dz_zt,wtracer_sec,&        ! Input
+         rho_zt,zt_grid,zi_grid,&         ! Input
+         tracer)                          ! Input/Output
+! INPUT VARIABLES
+  ! number of SHOC columns
+  integer, intent(in) :: shcol
+  ! number of vertical levels
+  integer, intent(in) :: nlev
+  ! number of interface levels
+  integer, intent(in) :: nlevi
+  ! number of tracers
+  integer, intent(in) :: num_tracer
+  ! time step [s]
+  real(rtype), intent(in) :: dtime
+  ! thickness of grid centered on thermo points [m]
+  real(rtype), intent(in) :: dz_zt(shcol,nlev)
+  ! vertical flux of tracers [varies]
+  real(rtype), intent(in) :: wtracer_sec(shcol,nlevi,num_tracer)
+  ! air density [kg/m3]
+  real(rtype), intent(in) :: rho_zt(shcol,nlev)
+  ! heights centered on thermo points [m]
+  real(rtype), intent(in) :: zt_grid(shcol,nlev)
+  ! heights centered on interface points [m]
+  real(rtype), intent(in) :: zi_grid(shcol,nlevi)
+
+! IN/OUT VARIABLES
+  ! tracers [varies]
+  real(rtype), intent(inout) :: tracer(shcol,nlev,num_tracer)
+
+! LOCAL VARIABLES
+  integer :: kb, kt, k, i, p
+  real(rtype) :: thedz, r1, r2, r3
+  real(rtype) :: rho_zi(shcol,nlevi)
+
+  ! linearly interpolate air density from thermo to interface grid
+  call linear_interp(zt_grid,zi_grid,rho_zt,rho_zi,nlev,nlevi,shcol,0._rtype)
+
+  do i=1,shcol
+    do k=1,nlev
+      kb=k+1
```

Comment:
@xyuan : Unless there is a strong reason for maintaining this looping order, we should switch the ordering and do something like:
```
    do k=1,nlev
      kb=k+1
      do i=1,shcol
```
This ordering will be faster in Fortran.